### PR TITLE
Add tests for BC7 and UASTC4x4 transcoding.

### PR DESCRIPTION
**Pull request by donmccurdy**
Output:

Screen Shot 2020-09-15 at 6 56 28 PM

### CODE REVIEW

**Comment by lexaknyazev** [comment]
Looks great! Couple suggestions.
- Replace single avocado image with the generator (#9)
- Add CI via GitHub Actions:
  - ✔️ `npm run asbuild` succeeds
  - ✔️ `build/` directory doesn't change after running `npm run asbuild`
  - ✔️ BC7, ASTC, and RGBA32_UNORM transcoding tests

Note that the output of our BC7 transcoder for UASTC Mode Index 8 blocks doesn't match the reference (MSC). This doesn't affect texture sampling results. We may need to maintain a custom MSC build to include those blocks in CI.

**Comment by donmccurdy** [comment]
How should the first change be structured? I'm assuming:

1. Generate 'source' texture with the generator offline.
2. Generate 'golden' transcoder output offline (with this tool or the MSC transcoder).
3. Check the 'source' and 'golden' files into version control.
4. When running the unit tests, the 'source' is transcoded and compared against the 'golden' output. 

Is that what you had in mind? I don't think that the golden files need to come from the MSC tool necessarily. We can generate goldens with these transcoders, keep them checked into the repository, and manually inspect output images if the output is changed intentionally — I assume that is rare.

**Comment by lexaknyazev** [comment]
UASTC is a very complex format. I wouldn't expect any "git-friendly" amount of test data to cover all possible transcoder code paths. For example, the latest ASTC transcoder bug (fixed earlier today) was affecting only 0.12% of 2.5GB dataset.

**Comment by donmccurdy** [comment]
I see... in that case we will need the reference transcoder to be published on NPM to use it in our tests. I'm not sure when that will happen, since we're still waiting for various changes to be merged to the basis_universal repository. I don't know whether the current MSC transcoder works in node.js at all, and wouldn't want to maintain another variant of it.

Would you be okay with a few simple (not exhaustive) tests checked into Git in the meantime, to unblock https://github.com/KhronosGroup/Universal-Texture-Transcoders/issues/7#issuecomment-688376393?

**Comment by lexaknyazev** [comment]
Here's my current local setup:
- 2.5 GB of random data created with the generator + 320 MB converted from the sample models repo.
- A reference (MSC) transcoder build with tweaked UASTC Mode Index 8 to BC7 transcoding.
- Everything runs in Node.js CLI.

I don't know much about npm best practices, but the following setup seems pretty straightforward:
- Compile a minimal (only needed formats) reference transcoder with the required BC7 tweak and check it in here (I can provide and maintain that easily).
- During the CI run, generate and transcode data in memory, compare the results.

**Comment by donmccurdy** [comment]
Checking in the modified reference transcoder might be good for now, then. How long does the generator take to generate 2.5 GB of sample data? I'd probably want to leave the glTF-Sample-Models repository out of the CI tests, downloading and encoding those would make CI much slower and potentially fragile.

**Comment by lexaknyazev** [comment]
The generator completes 19 modes * 8388608 blocks * 16 bytes in 12 seconds, including disk IO. We could speed up this by requesting fewer blocks for more simple modes (out of scope for now).

**Comment by donmccurdy** [comment]
Ok, that seems more than sufficiently quickly. 🙂 

I'll leave this PR alone for now, and update it after a reference transcoder is available, then?

**Comment by lexaknyazev** [comment]
With the generator and the MSC transcoder merged, I think we can proceed here.

**Comment by donmccurdy** [comment]
@lexaknyazev I've integrated the generator and MSC transcoder. A few notes:

Transcoding 1024x1024 textures, it took about 3 minutes to complete tests for the two transcode targets above. That seemed slower than I'd like for quickly testing a change. So I included two macros here: `npm run test` does a faster build on 128x128 textures and completes in 12 seconds. `npm run test:ci` does a full test on 1024x1024 textures, and is intended to run on the GitHub Action (or manually, when needed).

Does this look OK so far? I'll add RGBA32_UNORM if so.

I'd like to check in the tests first, and then set up the CI in a separate PR.

File: test/reference.ts
Comment (lexaknyazev) [inline test/reference.ts:17]: The generator usage example uses `8 * 1024 * 1024` as a "more readable" way of expressing 2<sup>23</sup> blocks. This corresponds to, for example, 2048x4096 (8192x16384 px) or 1x8388608 (4x33554432 px) texture blocks.

To avoid potential misunderstanding, let's use something more canonical here.

File: test/test.ts
Comment (lexaknyazev) [inline test/test.ts:61]: To deduplicate code with ASTC4x4, could this be refactored into a function that accepts module and target format pair?

File: test/test.ts
Comment (lexaknyazev) [inline test/test.ts:76]: Given that the transcoded data length is always a multiple of 16, it's likely faster to create `Uint32Array` views and compare 32-bit values.

File: test/test.ts
Comment (lexaknyazev) [inline test/test.ts:7]: Ditto.

**Comment by lexaknyazev** [comment]
Looks great overall, added a few comments.

As a future CI performance enhancement, we should be able to determine the minimum number of generated blocks for each of 19 modes that covers all possible transcode code paths.

File: test/reference.ts
Comment (donmccurdy) [reply test/reference.ts:17]: I'm not sure what you mean by more canonical, but I'll just omit the default value since it's passed into the constructor anyway.

File: test/test.ts
Comment (donmccurdy) [reply test/test.ts:7]: I'm not sure what you mean by 'more canonical', could you clarify? Note that the RGBA32 decoder example code deals with width/height explicitly when computing uncompressed length.

File: test/test.ts
Comment (donmccurdy) [inline test/test.ts:10]: The docs for RGBA32 show two lines that look contradictory:

```js
const uncompressedByteLength = width * yBlocks * 4 * 4;
```

```js
const textureByteLength = width * height * 4;
let decodedTextureView = new Uint8Array(memory.buffer, 65536 + compressedByteLength, textureByteLength);
```

Shouldn't `uncompressedByteLength` and `textureByteLength` match?

**Comment by donmccurdy** [comment]
Added the RGBA32 test — it's failing, with:

```
Incorrect array length. Expected 33554432 but got 262144.
```

With 256 x 256 x 4 = 262144, the reference transcoder is returning a larger response than I expect here. 😕

File: test/test.ts
Comment (donmccurdy) [reply test/test.ts]: Simplified this a bit — I think it's short enough now to not need a helper function.

File: test/test.ts
Comment (lexaknyazev) [reply test/test.ts:10]: These are correct. When the texture's height is a multiple of 4, these lines evaluate to the same value. In other cases, `uncompressedByteLength` would be greater than `textureByteLength` because the decoder uses that memory anyway.

File: test/test.ts
Comment (lexaknyazev) [reply test/test.ts]: `8` is an arbitrary number here. If we need to derive `nBlocks` from the texture's size, it should be like (to keep the value the same):
```
const width = 8192;
const height = 16384;
const nBlocks = ((width + 3) >> 2) * ((height + 3) >> 2);
```
In this case, it also would be better to use `width` and `height` in `ImageInfo` for the reference transcoder instead of hard-coded `4` and `4 * nBlocks`,

Comment (donmccurdy) [reply test/test.ts]: Ok, yeah that's clearer, and it ironed out the test failure I'd been getting with RGBA32. 👍

File: test/test.ts
Comment (donmccurdy) [reply test/test.ts]: Maybe I was over testing – generating a 1024x1024 image for each mode the tests are more than fast enough, so probably no need to change the view type.

**Comment by donmccurdy** [comment]
Tests are passing. ✅

**Comment by lexaknyazev** [comment]
Would it be possible to add an extra RGBA32 test for images with non-multiples-of-4 dimensions (like `(WIDTH - 1) * (HEIGHT - 1)`?

File: test/test.ts
Comment (donmccurdy) [inline test/test.ts:85]: I would've expected this to require `WIDTH - 1`  and `HEIGHT - 1`, but it passes as-is, and not with the change. Any idea why?

**Comment by donmccurdy** [comment]
OK — the `RGBA32 - npot` test passes, but I'm not sure I understand part of it.

File: test/test.ts
Comment (lexaknyazev) [inline test/test.ts:20]: Shouldn't this line use local variables instead of top-level consts?

File: test/test.ts
Comment (lexaknyazev) [inline test/test.ts:32]: Same here.

File: test/test.ts
Comment (donmccurdy) [reply test/test.ts:85]: Hang on, that's a typo...

**Comment by donmccurdy** [comment]
You caught the main issue here, but with variable scopes fixed there's still a failure on:

```
Incorrect array length. Expected 4186116 but got 4190208.
```

File: test/test.ts
Comment (lexaknyazev) [inline test/test.ts:86]: `uncompressedView` byte length is greater than the actual texture's byte length for non-multiple-of-4 height.

File: test/test.ts
Comment (lexaknyazev) [inline test/test.ts:79]: Maybe use more precise name here as "npot" is more strict than "non-multiple-of-4"?

**Comment by donmccurdy** [comment]
Looks good now: https://github.com/donmccurdy/Universal-Texture-Transcoders/actions/runs/260538042

File: .github/workflows/ci.yml
Comment (lexaknyazev) [inline .github/workflows/ci.yml:21]: Could we check here that rebuilding doesn't change the `build/` directory?

File: package.json
Comment (lexaknyazev) [inline package.json:37]: This command doesn't work on Windows (it needs to be like `set CI=1`). If there's no easy way to make this portable, we could keep it as is.

File: .github/workflows/ci.yml
Comment (donmccurdy) [reply .github/workflows/ci.yml:21]: Are you sure we want that enforced on pull requests? Typically I would ask external contributors to omit a `build/` directory from their PRs for easier review, and rebuild it myself (e.g. in a release script). Maybe this could be left for a later addition?

File: package.json
Comment (donmccurdy) [reply package.json]: Good to know. The CI doesn't actually need this macro — its env variable is configured by the `.yml` file — so let's just omit it and we can use the OS-specific commands as needed.

File: .github/workflows/ci.yml
Comment (lexaknyazev) [reply .github/workflows/ci.yml:21]: As we have `build/` directory checked-in, I'd like to have some way to ensure that the `main` branch is always in sync with itself. This doesn't have to happen on pull requests, so let's add this later.

**Comment by lexaknyazev** [comment]
Since this PR adds new deps, please commit the updated `package-lock.json`.

**Comment by donmccurdy** [comment]
Done 👍

**Review by lexaknyazev — APPROVED** [review]

**Pull request merged**
